### DISCUSSION TITLE
How to build both C and C++ binaries in the same meson.build file

### DISCUSSION THREAD

**Discussion by ericcurtin** [mesonbuild/meson#11788] [Q&A]
I opened this MR:

https://gitlab.com/sdm845-mainline/qbootctl/-/merge_requests/5

which builds a binary with the C++ library included and one without it included, but it uses C++ compilers and linkers for both binaries. The only reason this works is I manually built the C one with a C compiler and when using this build file the C++ library is optimized out when there are no references to it. It would be very easy for C++ to sneak into that binary again without the build failing.

Is there a way of making sure the C binary builds with only a C toolchain? Compiler, linker, etc. And that the C++ builds with a C++ toolchain?

**Comment by eli-schwartz** [comment]
The compiler used for *any* source file, regardless of whatever other compilers are available in the build context, should always be the correct compiler for that source filetype.

So, .cpp files will always be *compiled* with the CXX compiler, and .c files will always be *compiled* with the C compiler. The question is which linker to use... meson will try to autodetect the optimal linker to use based on the sources included in that target, but you can also control this with `link_language: <str>`, see https://mesonbuild.com/Reference-manual_functions.html#executable_link_language

**Reply by ericcurtin** [reply]
> So, .cpp files will always be compiled with the CXX compiler, and .c files will always be compiled with the C compiler.

So this is a problem for me. Because I want to use the same files for C source code and C++ source code. And changing the linker is not enough I've found, as there are some compile time differences between valid C and valid C++ also.

**Reply by eli-schwartz** [reply]
One option might be:
```meson
fs = import('fs')

cpp_src = fs.copyfile('foo.c', 'foo.cpp')
```

This causes meson -- and the compiler -- to see files named with the appropriate extension based on how you want to compile it.

Another option might be to compile with the CXX compiler but pass the `-x c` option via cpp_args to tell it to compile in C mode. Meson has no understanding of this and cannot predict how it will behave, so I recommend using the copyfile function...
